### DISCUSSION TITLE
Hello, Maven!

### DISCUSSION THREAD

**Discussion by raupachz** [support-and-care/maven-support-and-care#30] [Ideas for Funding]
Maven and Java development goes hand in hand. Maven is an essential tool for developing software with Java. If you are new to Java, you will probably encounter Maven very early in your learning path. Programming languages use "Hello, world!" to introduce themselves to a new audience. Build systems (I know Maven can do more) should use the same approach. I would like to suggest an easier way to create new projects from scratch. 

This is the [current way](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html) to create a project with Maven:

```
mvn archetype:generate -DgroupId=com.mycompany.app -DartifactId=my-app -DarchetypeArtifactId=maven-archetype-quickstart -DarchetypeVersion=1.4 -DinteractiveMode=false
```

This command is not intuitive and takes quite a few parameters. A beginner with Maven has no idea what all these parameters are. I doubt even an experienced Maven user knows the parameters by heart.

The output after running the command:

```
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------< org.apache.maven:standalone-pom >-------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] >>> archetype:3.2.1:generate (default-cli) > generate-sources @ standalone-pom >>>
[INFO] 
[INFO] <<< archetype:3.2.1:generate (default-cli) < generate-sources @ standalone-pom <<<
[INFO] 
[INFO] 
[INFO] --- archetype:3.2.1:generate (default-cli) @ standalone-pom ---
[INFO] Generating project in Batch mode
[INFO] ----------------------------------------------------------------------------
[INFO] Using following parameters for creating project from Archetype: maven-archetype-quickstart:1.4
[INFO] ----------------------------------------------------------------------------
[INFO] Parameter: groupId, Value: com.mycompany.app
[INFO] Parameter: artifactId, Value: my-app
[INFO] Parameter: version, Value: 1.0-SNAPSHOT
[INFO] Parameter: package, Value: com.mycompany.app
[INFO] Parameter: packageInPathFormat, Value: com/mycompany/app
[INFO] Parameter: package, Value: com.mycompany.app
[INFO] Parameter: groupId, Value: com.mycompany.app
[INFO] Parameter: artifactId, Value: my-app
[INFO] Parameter: version, Value: 1.0-SNAPSHOT
[INFO] Project created from Archetype in dir: /Users/raupach/Development/my-app
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.745 s
[INFO] Finished at: 2024-05-19T11:27:47+02:00
[INFO] ------------------------------------------------------------------------
```

Again, new Maven user won't be able to make any sense of this. It is a lot of information to take in. I would argue that the verbosity is more confusing than helpful. For example it says: BUILD SUCCESS. What did it build? All I wanted to do was create a project. Side note: if this is the first time Maven downloads additional dependencies.

My last point: I doubt the project is even usable. It is full of outdated dependencies. JUnit 4.11, Maven Compiler Plugin 3.0.8 and so fourth.

Things to improve on:
* Reduce verbosity 
* Fewer input arguments
* Everything included: no need to download additional dependencies
* Up to date Maven plugins and dependencies

Examples from other programming languages / build tools:

Rust and Cargo
```
cargo new my-app
cd my-app
```

Node and npm
```
mkdir my-app
cd my-app
npm init
```

Go
```
mkdir my-app
cd my-app
go mod init example/hello
```

**Comment by hendrikebbers** [comment]
I was already thinking about a website like https://start.spring.io or https://start.microprofile.io That goes in a similar direction.

**Reply by raupachz** [reply]
Yes. That would be nice. A good start in the right direction. However I still think that should be part of Maven itself not a webservice.

**Reply by khmarbaise** [reply]
The spring boot starter itself is an open source project... I did already some tries with it... need more time for that... I could think of `start.maven.org` ??? 
Also a more convenient command line application could be also a good enhancement/improvement...

**Reply by rmannibucau** [reply]
Does it solve any issue except having something fun in demo?
AFAIK IDE already integrate archectypes not that bad and cli is generally as pointless since using any editor is sufficient these days.
So until the starter does handle dependencies - and here we enter into the opiniated side of this kind of solution - I'm not sure it is worth it.

**Reply by laeubi** [reply]
> AFAIK IDE already integrate archectypes not that bad and cli is generally as pointless since using any editor is sufficient these days.

m2e allows to generate a maven "simple" project without even archetype. I don't think its actually a point generate a really "simple" project because I would not need maven / any tool for spit out a bunch of text lines.

Archetypes come into play when you really have a complex / boilerplate (e.g. webapp maybe spring and alike) and want to allow people to get this customized a bit (e.g. name of the webapp, used version of the lib and so on) to create the inital directory structure. It does not feel anything useful for something I can copy and paste from a few text boxes in a web-tutorial!

**Comment by raupachz** [comment]
@rmannibucau Let me explain:

The current process is broken. Maven out of the box doesn't provide an easy, coherent and correct way to start a project. A set of blessed and maintained archetypes might help, but I don't know of any at the moment.

Archetypes add complexity. I am trying to remove complexity. 

Maven is a CLI tool. I do a lot of terminal work. I can do everything: source control, building, reporting. The very basic thing I can't do is create a project. I would like to have an easy way to do that.

Yes, the IDE already provides a lot of things. But I don't want to tie my build lifecycle to an IDE. We have been there. Maven was the way out.

Lastly, everyting is opiniated. Maven is an opiniated tool. It came with a standard directory layout for example.

**Reply by rmannibucau** [reply]
My 2cts are that if IDE are not helping enough then, in CLI, echo does and thzre is no complexity there - archetypes are harder than "vi-ing" a pom IMHO so kot sure how to help there except explaining archetypes are one option only.

**Comment by laeubi** [comment]
> * Fewer input arguments

The example is more an "unattended install", the minimum example would be `mvn archetype:generate -DarchetypeArtifactId=maven-archetype-quickstart`

**Reply by raupachz** [reply]
@laeubi The minimum example would be `mvn init` or `mvn create webapp`. Without downloading the whole world. Everything else can be changed afterwards.

**Reply by rmannibucau** [reply]
Some questions:

* How différent is it from an archetype?
* We likely do not want to hardcode some pointless defaults (touch pom && mkdirs)
* It is likely it stays in a plugin (if downloads are an issue archetype can be slimmed down to lower the requirements a lot)
* Init vs webapp, already makes 2 archetypes so sounds like you just want an archetype ng plugin with a more powerful templating integration solution (types)

If we review the issues:

Things to improve on:

> Reduce verbosity

If it a taste matter but already doable in your home

Fewer input arguments

Dépends the archetype

Everything included: no need to download additional dependencies

Unlikely by design until we make default plug-ins part of maven

Up to date Maven plugins and dependencies

Same, doesn't mean anything anymore in maven 4 until you use central as a référence and have not Idea if your archetype works

So long story short this issue is maybe worth just investing in archetype plugin

**Reply by raupachz** [reply]
> How différent is it from an archetype?

Doesn't matter from a user perspective. Technical details. I a not a Maven contributor, less a plugin developer and I think I tried working with archetypes once or twice in the last twenty years. So I cannot answer that question. But if archetypes work, which they do, then maybe this is build around that? Some shell script magic? Or a new lifecycle that binds to some preconfigured maven archetype? Not sure if that works, though. 

> We likely do not want to hardcode some pointless defaults (touch pom && mkdirs)

Then call it _intelligent defaults_ like a standard directory.

> It is likely it stays in a plugin (if downloads are an issue archetype can be slimmed down to lower the requirements a lot)
> Init vs webapp, already makes 2 archetypes so sounds like you just want an archetype ng plugin with a more powerful templating integration solution (types)

Can't answer that as your average Joe user. Not a Maven contributor.

> Unlikely by design until we make default plug-ins part of maven

I don't understand why this is not the default. Why do I have to set versions for core stuff? Like maven compiler plugin. It is great that I can do that, but I prefer things included. Download the distribution, set the path and work the cli tool.

**Reply by rmannibucau** [reply]
> Then call it intelligent defaults like a standard directory.

While I like the intent and idea, I have no idea what it means.
A lib will have pom.xml(packaging=jar) + src/main/java (what all artifacts do), a webapp will have the exact same with packaging=war and be it (src/main/webapp is totally optional and no more an intelligent default) and to be honest war are kind of "past" these days so creating a webapp is creating a jar if we want to be "intelligent".

Overall my point is there is either too much cases or the defaults are not helping enough - we went through that multiple times.

Archetype enables you to select your template which helps a bit even if not magic.

What can be is to set in your settings.xml the defaults so you do not have to type them.

> Why do I have to set versions for core stuff?

this is a long topic and I'm part of the ones not willing it but long story short, if you do not you are locked to a maven version (any other version is not guaranteed to build even if it can).
I'm still hoping it is "reverted" but this is another issue this one should just embraces what is the "current" practise IMHO.

> It is great that I can do that, but I prefer things included. Download the distribution, set the path and work the cli tool.

Exactly the issue, it is just not the case until you do not define the plugin at all which makes the build fragile.
But it is a nice proposal: do not log a warn if no configuration at all is set, even using parents.
Can maybe help to recreate the issue about "no explicit plugin" need.

**Comment by laeubi** [comment]
I must confess I don't really get the intend here. You wrote you want a "Hello world" and reference a docu that exactly do that, only copy paste and it works (lets forget about the **actual outcome** could be improved of course, what would be only a matter of either improve the archetype or create a new one).

Even if it would be quite easy to change the script that ships with maven to detect a `mvn init` and under the hood call the above command I really question its value when it comes to your second one "learning maven", so _what would a user really "learn" from that_? Nothing!

Instead after I have played with this Hello World a little bit, I might notice that `com.mycompany.app` is not that nice as a groupid and probably "my-app" is a bad choice (again we might argue to better change it to `hello-maven` or similar), so I probably would investigate the commandline and draw some conclusions and then _actually learn something how maven works_! Like I do by modify any "Hello World" program to do something more in the direction of my goals.

Next about the "I don't want it to download things", this sounds counterproductive to me from many perspectives.

1. This is how maven works, so if I want to learn maven I should probably get familiar with that aspect as soon as possible and I might next discover things like  `dependency:go-offline` and alike
2. A download ensures that maven actually *can* download things, while this sounds obvious in the first place, there are many configurations where one e.g. you want to configure a company mirror, or there is a proxy or firewall and so on. So how does an offline example help you if you will fail the first time you add a real dependency? If I have no dependency why would I need maven?
3. The benefit of that is that always the latest examples can be used, and I can use a myriad of other things. No need to download maven again just because the example project changes.
4. How far should we go and how big should we grow the distribution? And for what value? Even if we would include just the things for this call it will bloat the distribution with something that is outdated and never used in real builds.

**Reply by raupachz** [reply]
@laeubi  Good discussion. Let me elaborate.

> "learning maven", so what would a user really "learn" from that? Nothing!

The intent is not to learn right now, but to make things simpler for beginners. It is discouraging for beginners if you start with complicated stuff right from the start even it is just a simple line that you and me can just copy&paste in a terminal window.

```
mvn init
```

is much less intimidating than

```
mvn archetype:generate -DgroupId=com.mycompany.app -DartifactId=my-app -DarchetypeArtifactId=maven-archetype-quickstart -DarchetypeVersion=1.4 -DinteractiveMode=false
```

the success of frameworks, tools and programming languages depends, among other thins, on how fast and easily you can attract new users. New users like simple things. Not complex stuff. That is why we are having Compact Source Files (JEP 512) and Instance Main Methods in Java. It is the same line of reasoning. That is why Spring Boot, Quarkus and other frameworks use their own easy to use tools to setup projects. They know, if they make it too hard to start right away, they won't attract new users and loose traction.

> Next about the "I don't want it to download things", this sounds counterproductive to me from many perspectives.

That is not what I said. Quote: Everything included: no need to download additional dependencies. I am talking about the basic builtin staff. Like some blessed plugins or something. Downloading dependencies, referencing them and working offline is great. Look at `pnpm` and `npm`. Maven is years ahead of them. They are now solving problems, that Maven solved years ago.

> How far should we go and how big should we grow the distribution? And for what value? Even if we would include just the things for this call it will bloat the distribution with something that is outdated and never used in real builds.

`apache-maven-3.9.11-bin.zip` is 9 MB. To make it work I need to download a JVM bringing in hundreds of MB. Bloat is not an issue. Maybe some bulletin blessed plugins that are included in the distribution and are updated regularly with every update? Think of a `.m2` repository within the distribution directory. I know, far fetched but this is a discussion forum. No harm done :)

**Reply by rmannibucau** [reply]
> mvn init

can you define what it does?

npm will skaffold a package.json and an empty index.js....we can skaffold an empty pom.xml which is literally a few lines (project+gav)

not sure how it would help but it can be an `echo > pom.xml` quite easily - and you can copy/paste it, even in a shell only machine.

now the challenge is that once you got that command...you need to change everything to get something working (packaging if you want a webapp, some dependencies, values most of the time - keep in mind npm is `$name` and we are `$gav+packaging` already.

So archetype or a (java now :)) script is likely the best you'll get if you want to stay useful for somebody IMHO.

**Comment by laeubi** [comment]
By the way I started a little experiment now, lets assume I want a simple maven project to learn from and I create a new repository and simply ask copilote to create the content for me:

<img width="1039" height="1622" alt="grafik" src="https://github.com/user-attachments/assets/3421cced-c1ee-4662-ab4d-1b5c7e355066" />

This is the resulting plan / PR from copilote:

https://github.com/laeubi/maven-starter/pull/1


So if we say

> This is the [current way](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html) to create a project with Maven

it probably is more "This was previously the way" but today one just ask the AI and static "starter pages" become obsolete before we have created one ;-)

Of course I could have asked also to create a simple web app or whatever else, so just some thought:

- No need to remember "complicated" command-line
- No verbose output
- Everything included / no additional dependencies (even no need to install maven)

**Reply by rmannibucau** [reply]
I like to add that option but from my XP it has a few drawbacks:

* Often outdated
* Limited to mainstream tech (to générateur a few content like tests compared to Dynamic starters)
* Needs some $$ in some context
* Needs to be allowed by companies (#policies)

**Reply by laeubi** [reply]
> Often outdated

You mean like maven archetypes / docu / websites ;-)

> Limited to mainstream tech (to générateur a few content like tests compared to Dynamic starters)

I though maven where mainstream enough isn't it? Also I though we are really talking about simple maven Hello World and not full blown complex applications to be provided out-of-the-box?

> Needs some $$ in some context

Everything needs to be paid in some way or the other
<img width="597" height="555" alt="grafik" src="https://github.com/user-attachments/assets/afcb2650-79ba-47a1-a26a-d47aa5fd0123" />

> Needs to be allowed by companies (#policies)

Whats the difference between a "smart starter website" and a "smart AI", its all mechanics.
And even looking at stack-overflow, copy random examples from website or books *might* be subject to the same concerns.

Also the inital request was:

> If you are new to Java, you will probably encounter Maven very early in your learning path. 

So I would not expect anything from this to be part of a commercial product anyways.

**Reply by rmannibucau** [reply]
Outdated: starter can be dynamic like talend one and are managed by you, not by an opaque learning strategy
Maven is mainstream but as mentionned it is pointless alone so you always need more...and today we want maven 4 which is  known by non ia properly
Paid: right but ia is not today
Smart starter vs ia: done right by design vs random (hallucinations and friends). All ia ARE commercial today, you expire free rate very quickly if intégrateur in an ide for ex

**Reply by laeubi** [reply]
> Maven is mainstream but as mentionned it is pointless alone so you always need more...

The I don't understand the intend of this "Hello Maven" thing .. I though it was about a very bare basic hello world.

>and today we want maven 4 which is known by non ia properly

Lets see what happens:
- https://github.com/laeubi/maven-starter/pull/2

> Smart starter vs ia: done right by design vs random (hallucinations and friends)

There is so much "wrong" things provided an claimed to be "right" I think everyone first needs to learn to distinguish between before using any "starter"

>  if intégrateur in an ide for ex

A starter page would have no IDE integration, so I tried to find something that is similar (e.g. a webpage, some simple commands, a result I can look at).

In any case I just wanted to bring up that aspect here as its quite fun to experiment with whats already possible with AI... and it often show we overestimate our own "creativity amount" when it comes to such things like "providing the right example fro beginners" it more tried to become a quite fixed template a tool could can came up with equally well.

**Reply by rmannibucau** [reply]
Once again I think there is space for IA for some people but for 80% of dev a remote script would give more flexibility (curl | sh or powsh flavor for win) and solve almost all issues without having anything else than a static site...but not sure how it fits maven, fits more some projects like spring, spark, ...
